internal/xfn/cached/cached_runner.go

```diff
@@ -0,0 +1,387 @@
+// for an identical request with an unexpired TTL.
+type FileBackedRunner struct {
+	wrapped FunctionRunner
+	fs      afero.Afero
+	log     logging.Logger
+	metrics Metrics
+}
+
+// A FileBackedRunnerOption configures a FileBackedRunner.
+type FileBackedRunnerOption func(r *FileBackedRunner)
+
+// WithLogger specifies which logger the FileBackedRunner should use.
+func WithLogger(l logging.Logger) FileBackedRunnerOption {
+	return func(r *FileBackedRunner) {
+		r.log = l
+	}
+}
+
+// WithMetrics specifies which metrics the FileBackedRunner should use.
+func WithMetrics(m Metrics) FileBackedRunnerOption {
+	return func(r *FileBackedRunner) {
+		r.metrics = m
+	}
+}
+
+// WithFilesystem specifies which filesystem implementation the FileBackedRunner
+// should use. The runner will ignore its path argument and cache files at the
+// root of this filesystem. Wrap your desired filesystem with afero.BasePathFS
+// to cache files under a specific path.
+func WithFilesystem(fs afero.Fs) FileBackedRunnerOption {
+	return func(r *FileBackedRunner) {
+		r.fs = afero.Afero{Fs: fs}
+	}
+}
+
+// NewFileBackedRunner creates a new function runner that wraps another runner,
+// caching its responses to the filesystem.
+func NewFileBackedRunner(wrap FunctionRunner, path string, o ...FileBackedRunnerOption) *FileBackedRunner {
+	r := &FileBackedRunner{
+		wrapped: wrap,
+		fs:      afero.Afero{Fs: afero.NewBasePathFs(afero.NewOsFs(), path)},
+		log:     logging.NewNopLogger(),
+		metrics: &NopMetrics{},
+	}
+
+	for _, fn := range o {
+		fn(r)
+	}
+
+	return r
+}
+
+// RunFunction tries to return a response from cache. It falls back to calling
+// the wrapped runner.
+func (r *FileBackedRunner) RunFunction(ctx context.Context, name string, req *fnv1.RunFunctionRequest) (*fnv1.RunFunctionResponse, error) {
+	start := time.Now()
+	log := r.log.WithValues("name", name)
+
+	// If we don't have a cache key we can't perform a cache lookup, or
+	// cache the response. Just send it on. This should never happen.
+	if req.GetMeta().GetTag() == "" {
+		log.Debug("RunFunctionResponse cache miss", "reason", ReasonEmptyRequestTag)
+		r.metrics.Miss(name, ReasonEmptyRequestTag)
+		return r.wrapped.RunFunction(ctx, name, req)
+	}
+
+	key := filepath.Join(name, req.GetMeta().GetTag())
+	log = log.WithValues("cache-key", key)
+
+	f, err := r.fs.Open(key)
+	if errors.Is(err, fs.ErrNotExist) {
+		log.Debug("RunFunctionResponse cache miss", "reason", ReasonNotCached)
+		r.metrics.Miss(name, ReasonNotCached)
+		return r.CacheFunction(ctx, name, req)
+	}
+	if err != nil {
+		log.Info("RunFunctionResponse cache miss", "reason", ReasonError, "err", err)
+		r.metrics.Miss(name, ReasonError)
+		r.metrics.Error(name)
+		return r.CacheFunction(ctx, name, req)
+	}
+	defer f.Close() //nolint:errcheck // Only open for reading. Nothing useful to do with the error.
+
+	// The first 8 bytes of our file is always our deadline.
```

Comment:
Did you consider encapsulating the cache entry in a protobuf to simplify marshaling/unmarshaling? Something like:

```
message CachedFunctionResponse {
    google.protobuf.Timestamp deadline = 1;
    RunFunctionResponse response = 2;
}
```

Then the cache files would just be marshaled protobufs, absolving this code of the details of on-disk file format and making it easier to evolve that format in the future. I suppose this may make the cache a touch slower in the case where the deadline is expired, since we have to unmarshal the whole message.

Reply:
I did. I was reluctant in large part because today all of our code generation plumbing assumes all generated code lives under `apis/`, and I wouldn't want this cached response wrapper to live there. I'm also not too worried about the space or efficiency of deadline encoding in this scenario.

That said, I agree it'd be nice to not have to maintain this weird little encoding variant for this niche case. I'll take a pass at generating a `CachedFunctionResponse` and see how looks.

Reply:
@adamwg I pushed a commit that does this - PTAL.

I think it's an improvement. It'll use more memory during garbage collection but the code is simpler. It did require a large amount of build tooling changes, which you can see in several additional commits.